Training/Training/Reversal.c

```diff
@@ -0,0 +1,163 @@
+               break;
+            case '5':
+               return 0;
+               break;
+         }
+      } else printf (ANSI_COLOR_RED"Please enter number between 1-5\n\n"ANSI_RESET_ALL);
+   }
+   return 0;
+}
+
+int NumIsPalindrome (int num, int type) {
+   bool isNegative = false;
+   if (num < 0) {
+      num = -num;
+      isNegative = true;
+   }
+   int org = num;
+   long long int rev = 0;
+   while (num != 0) {
+      rev = rev * 10 + num % 10;
+      num /= 10;
+   }
+   if (type == 1) printf ("Output:%lld\n", rev);
+   if (isNegative == true) return NOT_PALINDROME;
+   return (rev == org);
+}
+
+int StrIsPalindrome (char* phrase, int type) {
+   int j = 0, length = 0;
+   length = (int)strlen (phrase);
+   if (type == 1) length--;
+   if (length > MAX_CHAR - 2) return INVALID;
+   char temp[MAX_CHAR] = { '0' };
+   for (int i = 0; i < length; i++) {
+      if (isalnum (phrase[i]) != 0)
+         temp[j++] = phrase[i];
+   }
+   temp[j] = '\0';
+   if (j == 0) return INVALID;
+   for (int n = 0; n < (j / 2); n++) {
+      if (toupper (temp[n]) != toupper (temp[j - n - 1])) return NOT_PALINDROME;
+   }
+   return PALINDROME;
+}
+
+void StringInput (int option) {
+   char ch = '\n';
+   char phrase[MAX_CHAR];
+   printf ("\nEnter a Phrase: ");
+   char* j = fgets (phrase, sizeof (phrase), stdin);
+   char* strResult = strchr (phrase, ch); //To check whether input exceeds limit
+   if (strResult == NULL) while (getchar () != '\n');
+   (j != NULL && phrase[0] != '\n' && strResult != NULL) ? PrintResult (StrIsPalindrome (phrase, 1)) : printf (ANSI_COLOR_RED"INVALID!!!\n\n"ANSI_RESET_ALL);
+}
+
+void IntegerInput (int option) {
+   char ch = '\n';
+   char input[MAX_DIGIT];
+   int number = 0;
+   char* endptr;
+   printf ("\nEnter an integer: ");
+   char* i = fgets (input, sizeof (input), stdin);
+   char* intResult = strchr (input, ch); //To check whether input exceeds limit
+   number = strtol (input, &endptr, 10);
+   if (intResult == NULL) while (getchar () != '\n');
+   (i == NULL || endptr == input || *endptr != '\n' || input[0] == ' ' || input[0] == '\t') ? printf (ANSI_COLOR_RED"INVALID!!!\n\n"ANSI_RESET_ALL) : (errno == ERANGE) ? printf (ANSI_COLOR_RED"INPUT LIMIT EXCEEDED!!!\n\n"ANSI_RESET_ALL) : PrintResult (NumIsPalindrome (number, 1));
+}
+
+void OutputCheck () {
+   printf ("\nPALINDROME CHECKER:\n\n--------------INPUT----------------------OUTPUT------------RESULT---");
+   char* strInput[] = { "Don't nod","I did, did I?","Was it a car or a cat I saw?","@@@","RADAR","NASA","CIVIC","NUN","TRUMPF METAMATION" };
+   int strLength = sizeof (strInput) / sizeof (strInput[0]);
+   for (int i = 0; i < strLength; i++) {
+      char* strOutput = "It is not a Palindrome";
+      char* strResult = ANSI_COLOR_RED"FAIL"ANSI_RESET_ALL;
+      if (StrIsPalindrome (strInput[i], 0) == 1 || StrIsPalindrome (strInput[i], 0) == 0 || StrIsPalindrome (strInput[i], 0) == -1) strResult = ANSI_COLOR_GREEN"PASS"ANSI_RESET_ALL;

```

Comment:
Store the result from the StrIsPalindrome function and use it to check. Avoid calling it several times to check. Apply this wherever possible.